### PR TITLE
Helm 3: initial registry support

### PR DESCRIPTION
**Pull request by jdolitsky**
This is initial support for OCI registries against the dev-v3 branch. It is not complete, and requires more UX design etc. Hoping we can get something merged soon and then iterate from there.

Here is a high-level summary of the changes:

### save

save a chart directory

```
$ helm chart save mychart/ localhost:5000/myrepo/mychart:latest
3344059: Saving meta (216 B)
84059d7: Saving content (454 B)
Name: mychart
Version: 2.7.0
Meta: sha256:3344059bb81c49cc6f2599a379da0a6c14313cf969f7b821aca18e489ba3991b
Content: sha256:84059d7403f496a1c63caf97fdc5e939ea39e561adbd98d0aa864d1b9fc9653f
latest: saved
```

### list

list all saved charts

```
$ helm chart list
REF                                                     NAME                    VERSION DIGEST  SIZE            CREATED
localhost:5000/myrepo/mychart:latest                    mychart                 2.7.1   84059d7 454 B           27 seconds
localhost:5000/stable/acs-engine-autoscaler:latest      acs-engine-autoscaler   2.2.2   d8d6762 4.3 KiB         2 hours
localhost:5000/stable/aerospike:latest                  aerospike               0.2.1   4aff638 3.7 KiB         2 hours
localhost:5000/stable/airflow:latest                    airflow                 0.13.0  c46cc43 28.1 KiB        2 hours
localhost:5000/stable/anchore-engine:latest             anchore-engine          0.10.0  3f3dcd7 34.3 KiB        2 hours
...
```

### export

export a chart to directory

```
$ helm chart export localhost:5000/myrepo/mychart:latest
Name: mychart
Version: 2.7.0
Meta: sha256:3344059bb81c49cc6f2599a379da0a6c14313cf969f7b821aca18e489ba3991b
Content: sha256:84059d7403f496a1c63caf97fdc5e939ea39e561adbd98d0aa864d1b9fc9653f
Exported to mychart/
```

### push

push a chart to remote

```
$ helm chart push localhost:5000/myrepo/mychart:latest
The push refers to repository [localhost:5000/myrepo/mychart]
Name: mychart
Version: 2.7.0
Meta: sha256:3344059bb81c49cc6f2599a379da0a6c14313cf969f7b821aca18e489ba3991b
Content: sha256:84059d7403f496a1c63caf97fdc5e939ea39e561adbd98d0aa864d1b9fc9653f
latest: pushed to remote (2 layers, 670 B total)
```

### pull

pull a chart from remote

```
$ helm chart pull localhost:5000/stable/wordpress:latest
latest: Pulling from localhost:5000/stable/wordpress
2c017c4: Saving meta (437 B)
8224586: Saving content (18.1 KiB)
Name: wordpress
Version: 5.1.2
Meta: sha256:2c017c46f229ef5faf021d54c2ca6df862169e4314ccdf324ee6faa23ebc585f
Content: sha256:8224586842c560dcbe3f98acd34aef243bb30233126af62efd3b2a82e4f3cae9
Status: Downloaded newer chart for localhost:5000/stable/wordpress:latest
```

### remove

```
$ helm chart remove localhost:5000/myrepo/mychart:latest
latest: removed
```

## Chart Manifest

This is the manifest of a Helm chart as stored in a registry:

```
$ curl -s -H 'Accept: application/vnd.oci.image.manifest.v1+json' http://localhost:5000/v2/myrepo/mychart/manifests/latest | jq
{
  "schemaVersion": 2,
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
    "size": 2
  },
  "layers": [
    {
      "mediaType": "application/vnd.cncf.helm.chart.meta.v1+json",
      "digest": "sha256:c356ec641a696eb5f3320bed9e8ceeb505fcc84b7ee072a85a8098fc362e13b7",
      "size": 210,
      "annotations": {
        "org.opencontainers.image.title": "chart-meta.json"
      }
    },
    {
      "mediaType": "application/vnd.cncf.helm.chart.content.v1+tar",
      "digest": "sha256:84059d7403f496a1c63caf97fdc5e939ea39e561adbd98d0aa864d1b9fc9653f",
      "size": 454,
      "annotations": {
        "sh.helm.chart.name": "mychart",
        "sh.helm.chart.version": "2.7.1",
        "org.opencontainers.image.title": "chart-content.tgz"
      }
    }
  ]
}
```

The manifest is built using [deislabs/oras](https://github.com/deislabs/oras).

## Chart Storage

Chart content is currently stored locally in `$HELM_HOME/registry`:

```
$ tree ~/.helm/registry
/Users/me/.helm/registry
├── blobs
│   └── sha256
│       ├── 3344059bb81c49cc6f2599a379da0a6c14313cf969f7b821aca18e489ba3991b
│       └── 84059d7403f496a1c63caf97fdc5e939ea39e561adbd98d0aa864d1b9fc9653f
├── charts
│   └── mychart
│       └── versions
│           └── 2.7.1
└── refs
    └── localhost_5000
        └── myrepo
            └── mychart
                └── tags
                    └── latest
                        ├── chart -> /Users/me/.helm/registry/charts/mychart/versions/2.7.1
                        ├── content -> /Users/me/.helm/registry/blobs/sha256/3344059bb81c49cc6f2599a379da0a6c14313cf969f7b821aca18e489ba3991b
                        └── meta -> /Users/me/.helm/registry/blobs/sha256/84059d7403f496a1c63caf97fdc5e939ea39e561adbd98d0aa864d1b9fc9653f
```

 Note: This is using a simple filesystem layout, but this is subject to change. We will potentially use something like [bbolt](https://github.com/etcd-io/bbolt).

### CODE REVIEW

File: pkg/registry/cache.go [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+	ch.Metadata.Name = ""
+	ch.Metadata.Version = ""
+	metaJSONRaw, err := json.Marshal(ch.Metadata)
+	if err != nil {
+		return nil, err
+	}
+	metaLayer := cache.store.Add(HelmChartMetaFileName, HelmChartMetaMediaType, metaJSONRaw)
+
+	// Create content layer
+	// TODO: something better than this hack. Currently needed for chartutil.Save()
```
Comment (adamreese) [inline pkg/registry/cache.go]: what is happening here?  can you add a comment about why this is required

File: cmd/helm/chart_list.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/helm/pkg/action"
+)
+
+const chartListDesc = `
+TODO
```
Comment (adamreese) [inline cmd/helm/chart_list.go]: Can you populate the descriptions?  It would help with understanding intent.

**Review by adamreese — CHANGES_REQUESTED** [review]
@jdolitsky I have posted my initial comments but still going through some of the workflow.

Please add links to any design/proposal docs.

File: pkg/registry/cache.go (hunk 1 above)
Comment (jdolitsky) [reply pkg/registry/cache.go]: Added some more comments in new commit, basically the Name and Version are required to save a chart. We need to rethink the relationship between metadata and chart content

File: pkg/action/chart_export.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+	ch, err := a.cfg.RegistryClient.LoadChart(r)
+	if err != nil {
+		return err
+	}
+
+	// Save the chart to local directory
+	// TODO: init in Helm home? Or no file creation at all?
+	tempDirPrefix := ".helm-chart-export"
+	os.MkdirAll(tempDirPrefix, 0755)
```
[the remark this answers is not in the document]
Comment (jdolitsky) [reply pkg/action/chart_export.go]: Yes, updated. Much cleaner using `chartutil.SaveDir`

File: cmd/helm/chart_list.go (hunk 2 above)
Comment (jdolitsky) [reply cmd/helm/chart_list.go]: done

**Comment by bacongobbler** [comment]
thanks @jdolitsky for taking this one on! Taking this for a spin now.

**Comment by bacongobbler** [comment]
Found a bug: regardless of the tag I provide, `:latest` is always appended to the end of the reference.

```
PS C:\Users\me\code\go\src\k8s.io\helm> .\bin\helm.exe create mychart
PS C:\Users\me\code\go\src\k8s.io\helm> .\bin\helm.exe chart save .\mychart\ mychart:0.1.0
Using default tag: latest
437a908: Saving meta (82 B)
5676733: Saving content (2.1 KiB)
Name: mychart
Version: 0.1.0
Meta: sha256:437a9084a8357638f098100ee5c3501eeafcdd72bb489dadfa6aa3f0cbaa8673
Content: sha256:5676733a1780522db2ae13bee4446b7327b76c89554a6552fef89a3987286db0
latest: saved
PS C:\Users\me\code\go\src\k8s.io\helm> .\bin\helm.exe chart list
REF                     NAME    VERSION DIGEST  SIZE    CREATED
mychart:0.1.0:latest    mychart 0.1.0   5676733 2.1 KiB About a minute
```

**Comment by bacongobbler** [comment]
And just to clarify: this only introduces the `helm chart` subcommands, but there's no way to install a chart that was `helm chart save`'d yet, right? I'm assuming that work has been bumped later down the line and we just want to get the repository interaction stuff in the core. Is that right?

**Comment by jdolitsky** [comment]
@bacongobbler I haven't had this play with `helm install` yet but it will work. The new architecture allows us to provide a shared `RegistryClient` that can be used across all subcommands, and methods are there to return a ref as a chart object.

Can you also try this command to debug the "latest" issue you described? :
```
.\bin\helm.exe chart save .\mychart\ localhost:5000/mychart:0.1.0
```
I'm thinking it is due to missing a host

**Comment by bacongobbler** [comment]
Ah! That did it. Thanks! For reference, the chart was saved correctly as `registries/refs/mychart_0.1.0` so I could interact with it as `mychart:0.1.0` regardless of the `helm chart list` output bug.

**Review by bacongobbler — APPROVED** [review]
Asides from the one UI bug, everything else works as intended. I was able to quickly stand up `registry:2` locally and push/pull charts to the registry, as well as save, list, and export charts to the OCI image layout. This is *awesome*. I really like where this work is going.

I'm gonna go ahead and LGTM this; the core business here is going in the right place. I'm in favour of merging this as-is and iterating from there.

File: pkg/registry/client_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+type RegistryClientTestSuite struct {
+	suite.Suite
+	Out                io.Writer
+	DockerRegistryHost string
+	RegistryClient     *Client
+}
+
+func (suite *RegistryClientTestSuite) SetupSuite() {
+	os.RemoveAll("helm-registry-test")
```
Comment (bacongobbler) [inline pkg/registry/client_test.go]: Would this be better in an `AfterTest()`? https://godoc.org/github.com/stretchr/testify/suite#AfterTest

That way we can get rid of the .gitignore file as well.

File: cmd/helm/chart_export.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/helm/cmd/helm/require"
+	"k8s.io/helm/pkg/action"
+)
+
+const chartExportDesc = `
+TODO
```
Comment (bacongobbler) [inline cmd/helm/chart_export.go]: If you have time, would you mind adding a small blurb on `helm chart export`?

File: pkg/action/chart_export.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+		return err
+	}
+
+	ch, err := a.cfg.RegistryClient.LoadChart(r)
+	if err != nil {
+		return err
+	}
+
+	// Save the chart to local directory
+	// TODO: make destination dir configurable
```
Comment (bacongobbler) [inline pkg/action/chart_export.go]: 👍 this was what I was going to ask next. Glad this is captured as a TODO item.

File: pkg/registry/client.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	table := uitable.New()
+	table.MaxColWidth = 60
+	table.AddRow("REF", "NAME", "VERSION", "DIGEST", "SIZE", "CREATED")
+	rows, err := c.cache.TableRows()
+	if err != nil {
+		return err
+	}
+	for _, row := range rows {
+		if len(row) == 6 {
+			table.AddRow(row[0], row[1], row[2], row[3], row[4], row[5])
```
Comment (bacongobbler) [inline pkg/registry/client.go]: I believe it's possible to do something like `table.AddRow(row...)`, which will iterate through each item in the list and add it as an argument.

here's an example in practice: https://play.golang.org/p/1Dq-DH6GZhc

**Comment by jzelinskie** [comment]
This looks pretty good so far. Agreed on merging this sooner rather than later and iterating on it.
One big thing that's probably missing is managing OCI Indexes (aka ManifestLists).

**Comment by sajayantony** [comment]
@jzelinskie  @jdolitsky - Could you elaborate the scenario why we would need to handle manifest lists?

**Comment by jzelinskie** [comment]
When pulling, you can get back either a single Manifest or ManifestList (aka OCI Index).
If you get the list, you have to resolve which one has the the relevant Manifest, pull that, and then pull all the layers.

The UX around pushing is pretty error-prone, you have to pull down a ManifestList, append your new manifest to it, and then push it. There is nothing preventing you from accidentally overwriting or racing with another client.

**Review by adamreese — APPROVED** [review]

File: pkg/registry/client.go (hunk 7 above)
Comment (bacongobbler) [reply pkg/registry/client.go]: @adamreese told me this isn't possible as `table.AddRow` accepts an `[]interface{}`. I had assumed wrongly that it accepted a string slice. My bad

File: cmd/helm/chart_export.go (hunk 5 above)
Comment (jdolitsky) [reply cmd/helm/chart_export.go]: done

File: pkg/registry/client_test.go (hunk 4 above)
Comment (jdolitsky) [reply pkg/registry/client_test.go]: done

File: pkg/registry/client.go (hunk 7 above)
Comment (jdolitsky) [reply pkg/registry/client.go]: made it happen :)

**Review by bacongobbler — APPROVED** [review]

**Comment by jdolitsky** [comment]
Merged. Thanks to all the people who helped this come together... blessed ✌️

I think it's a big step in an exciting direction for Helm that has in some form or another been under discussion on the project for (literally) years, started by our comrades at Quay/CoreOS.

Look forward to seeing the end result of these things in Helm 3!

**Pull request merged**